### PR TITLE
eth/filter: refactor log filter system

### PR DESCRIPTION
**Pull request by bas-vk**
This PR is a rewrite of the `eth/filter` package. It has a better separation between the different parts of the filter package which makes it better testable through unittests. The unittests are designed to run in parallel.

#### Structure
The API creates a filter manager and uses this manager to retrieve filter data. This makes the API in itself simple. The manager has 2 maps which contain both all filters. The `all` map which is a mapping from the filter id to a filter and an `index` which is used internally in the work loop and a mapping from its filter type, filter id to the filter. All calls from exported methods use the `all` map to access the filter. The internal work queue uses the index to access filters. Each filter has a channel, the work loop will write event data and the API will read filter data from this channel. In practice, each filter has 2 channels, one to store logs (log filters) and another to store hashes (block/tx filters).

On uninstall the filter is removed from the `all` map. This makes the filter invisible from the outside. Also an uninstall request is send to the work loop through the `uninstall` channel. When the work loop receives the request it will close the data channel and remove the filter from the index.

Each 30 seconds all filters are verified for activity. If a filter has not been used in the last 5 minutes it is considered inactive and uninstalled.

The data channels in the filter (`logs` and `hashes`) are buffered. When the channel is full the filter is considered inactive, removed and a warning is written to log.

The logs subscription is build on top of the filter system and is implemented in the API.

@fjl, PTAL

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Mon Jul 11 09:11:19 UTC 2016_

File: eth/filters/filter_system.go [hunk 1]
```diff
@@ -14,172 +14,588 @@
...
+}
+
+// Log is a helper that can hold additional information about vm.Log
+// necessary for the RPC interface.
+type Log struct {
+	*vm.Log
+	Removed bool `json:"removed"`
+}
+
+type filter struct {
```
Comment (fjl) [inline eth/filters/filter_system.go]: Hmm. Why is this a separate type, can't it just be `Filter`?

File: eth/filters/filter_system.go [hunk 2]
```diff
@@ -14,172 +14,588 @@
...
+
+	logHandler := func(f *filter, logs []Log) {
+		if f.lc != nil && len(logs) > 0 {
+			f.lc(f.ID, logs)
+		} else if len(logs) > 0 {
+			select {
+			case f.logs <- logs:
+				return
+			default: // data queue full, disable filter
+				inactive = append(inactive, f)
```
Comment (fjl) [inline eth/filters/filter_system.go]: I don't like this decision. I think it would be better to just block here instead.

Subscriptions are already buffered in package RPC and will be removed if the
buffer is full. We would need to ensure that `process` gets unblocked if the filter
is removed. One way to do that is to add an additional channel to the filter which
is closed when the filter is removed. You would then select on that channel here.

File: eth/filters/filter_system.go [hunk 3]
```diff
@@ -14,172 +14,588 @@
...
+// that are used to generate filter identifiers.
+func filterIDGenerator() *rand.Rand {
+	if seed, err := binary.ReadVarint(bufio.NewReader(crand.Reader)); err == nil {
+		return rand.New(rand.NewSource(seed))
+	}
+	return rand.New(rand.NewSource(int64(time.Now().Nanosecond())))
+}
+
+// newID generates a filter identifier.
+func newID() (FilterID, error) {
```
Comment (fjl) [inline eth/filters/filter_system.go]: Small nitpick, but please move the FilterID constructor stuff closer to the FilterID type.

File: eth/filters/filter_system_test.go [hunk 4]
```diff
@@ -17,101 +17,395 @@
...
+			// match logs based on addresses and topics
+			/* 4 */ {FilterCriteria{Addresses: []common.Address{thirdAddress}, Topics: [][]common.Hash{[]common.Hash{firstTopic, secondTopic}}}, append(concatLogs(allLogs[3:5]), allLogs[5].Logs[0]), FilterID{}},
+			// match logs based on multiple addresses and "or" topics
+			/* 5 */ {FilterCriteria{Addresses: []common.Address{secondAddr, thirdAddress}, Topics: [][]common.Hash{[]common.Hash{firstTopic, secondTopic}}}, append(concatLogs(allLogs[2:5]), allLogs[5].Logs[0]), FilterID{}},
+			// block numbers are ignored for filters created with New***Filter, these return all logs that match the given criterias when the state changes
+			/* 6 */ {FilterCriteria{Addresses: []common.Address{firstAddr}, FromBlock: rpc.BlockNumber(2), ToBlock: rpc.BlockNumber(3)}, append(concatLogs(allLogs[:2]), allLogs[5].Logs[3]), FilterID{}},
+			// multiple pending logs, should match only 2 topics from the logs in block 5
+			/* 7 */ {FilterCriteria{Addresses: []common.Address{thirdAddress}, Topics: [][]common.Hash{[]common.Hash{firstTopic, forthTopic}}}, vm.Logs{allLogs[5].Logs[0], allLogs[5].Logs[2]}, FilterID{}},
```
Comment (fjl) [inline eth/filters/filter_system_test.go]: Syntax Tip: you can write the slice indexes as actual indexes instead of putting them in as comments.

```go
		tests = []struct {
			crit     FilterCriteria
			expected vm.Logs
			id       FilterID
		}{
			// match all
			0: {FilterCriteria{}, concatLogs(allLogs), FilterID{}},
			// match none due to no matching addresses
			1: {FilterCriteria{Addresses: []common.Address{common.Address{}, notUsedAddress}, Topics: [][]common.Hash{[]common.Hash{}}}, vm.Logs{}, FilterID{}},
			...
```

File: eth/filters/filter_system.go (hunk 2 above)
Comment (bas-vk) [inline eth/filters/filter_system.go]: This is the case for subscriptions. But in general I think that the filter package should be self contained and not depend on an external package to uninstall the filter when the queue is full.

File: eth/filters/filter_system.go (hunk 1 above)
Comment (bas-vk) [inline eth/filters/filter_system.go]: The reason is that both do slightly different things, `Filter` is used to search for stored logs and `filter` is used by the filter system for state changes. We could combine them, but I prefer to do that in a separate PR.

File: eth/filters/filter_system.go (hunk 2 above)
Comment (fjl) [inline eth/filters/filter_system.go]: If you use filters from Go, it's fine to block. If the application cannot handle events fast enough, blocking creates backpressure that will eventually stop new events from being created until the application can keep up.

File: eth/filters/filter_system.go (hunk 1 above)
Comment (fjl) [inline eth/filters/filter_system.go]: ok

File: eth/filters/filter_system.go (hunk 2 above)
Comment (bas-vk) [inline eth/filters/filter_system.go]: The queue is used for polling, subscriptions use the callback which doesn't cause a problem. 

Consider the following case if we block or use an uninstall channel:
1. A DApp installs a block filter over http (polling)
2. The node start synchronising and generates a lot of these block events
3. The DApp doesn't query the filter for changes
4. The filter hash queue is full at some point, without the default case this will block until the filter is polled. We cannot expect the DApp to call the Uninstall method.

**Comment by fjl** [comment]
We had a call to talk about the blocking channel issue.
I have two goals in mind:

* There should only be one way in which filter updates are delivered.
  The filter manager API should either offer callbacks or send on a caller-supplied channel.
* Whatever Go API we come up with should mirror the subscription API provided by the RPC client.
  This is because a higher layer such as accounts/abi/bind/backend will need to work with both
  of them and it is super inconvenient to wrap two entirely different approaches.

For the channel-based approach, the API should probably look like this:

```go
func (m *Manager) NewLogFilter(crit FilterCriteria, chan []vm.Log) Subscription

// Note: this can also be a struct type containing the filter ID.
type Subscription interface {
    Unsubscribe() // removes the filter
}
```

One remaining question is how closing the subscription affects the channel.

* The channel can be closed. This is the approach currently taken by the RPC client.
  The rpc client makes the error available through the Subscription's `Err` method.

* The returned subscription object can contain an error channel that callers need to
   listen on. The error channel would receive a value when the filter is unsubscribed
   (by the user or the manager).

**Comment by bas-vk** [comment]
Closed in favor of: https://github.com/ethereum/go-ethereum/pull/2885
